### PR TITLE
Feat/talk2scholars zotero write

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors

## Zotero Write
https://github.com/user-attachments/assets/a5b8434c-f8ee-4eac-a574-a8128f0abe9a

#### I forgot to record the path feature in the first video, the video  below shows the nested path
https://github.com/user-attachments/assets/c036ec90-2d34-4a99-ac99-dae315383512

## Description

Introducing Zotero Write Tool, Enhancing Zotero Read, and Improving Error Handling  

#### **Key Updates in This PR**  

1. **Zotero Write Tool**:  
   - Added a new tool to **save papers to the user's Zotero library**.  
   - It retrieves papers from the `last_displayed_key` in the shared state and stores them in Zotero.  
   - Supports user workflows where a user searches for papers first and then decides to save them.  

2. **Enhancements to Zotero Read Tool**:  
   - Introduced **nested path support** to retrieve papers **along with their collection hierarchy**.  
   - Created a utility (`zotero_path`) to **recursively build collection paths** for both **Zotero Read** and **Zotero Write** tools.  
   - Now, when a user queries for papers (e.g., *"Find papers on machine learning in my Zotero library"*), the tool **retrieves and displays papers along with their full collection path** (e.g., `/temp/paper/cats`).  
   - This collection path is **stored in the shared state (`zotero_read` key)** for future reference.  

3. **Improved Error Handling for Zotero Write**:  
   - If a user **hasn't run the Zotero Read tool before saving**, the collection path won’t be available.  
   - To handle this, the tool **raises an error**, which triggers the `zotero_path` utility to **rebuild the collection path dynamically** before saving the paper.  
   - Added a **direct key-based matching** mechanism:  
     - If a user **misspells a collection name**, the tool **tries to match it with existing collections** before raising an error.  
     - If no match is found, the tool **suggests available collections** in the error message.  

4. **Fix for Semantic Scholar API Issues in S2 Agent**:  
   - Previously, if the API **failed to fetch results** due to connection issues, the LLM **misinterpreted the error** and changed the search term until it got the results from the api (e.g., from *"machine learning"* to *"applications of machine learning"*).  
   - **Solution**:  
     - Now, **all S2 tools raise explicit errors** when API calls fail.  
     - This ensures the LLM understands the failure as a **retrieval issue, not a lack of papers**. and triggers the same query and not chnging the name

5. **Migration to `create_supervisor` for the Main Agent**:  
   - Updated the main agent to use `create_supervisor`  

6. **Streaming Support in Streamlit**:  
   - Added **streaming responses** in the Streamlit app for a smoother user experience.
   - Added ArxivId fetch, and reduces system prompt to almost 10% of the original  

#### **Planned Improvements in Future PRs**  
- Further **refactoring of Zotero tools**:  
  - Move core logic from **Zotero Read** and **Zotero Write** tools into **a shared utility file** to reduce duplication. Similar to issue #107  

## Fixes # (issue) Mention the issue number.
#133  #112 #76 

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [x] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: aiagents4pharma/talk2scholars/tests/test_s2_search.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+def test_search_tool_unexpected_format(monkeypatch):
+    """
+    Test that search_tool raises a RuntimeError when the API response
+    does not include the expected 'data' key.
+    """
+    monkeypatch.setattr(requests, "get", dummy_requests_get_no_data)
+    tool_call_id = "test_tool_call_id"
+    with pytest.raises(
+        RuntimeError,
+        match="Unexpected response from Semantic Scholar API. Please retry the same query.",
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tests/test_s2_search.py:216]: If there is an error in the formatting (or no paper returned) should the agent retry the same query or would it better to let the user know and then modify their query accordingly?

File: aiagents4pharma/talk2scholars/tests/test_zotero_write.py [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+    def test_get_item_collections_exception(
+        self,
+        mock_get_item_collections,
+        mock_zotero_class,
+        mock_hydra_compose,
+        mock_hydra_init,
+    ):
+        """
+        Test that if get_item_collections raises an exception, the fallback branch
+        (lines 92-99) raises a RuntimeError.
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tests/test_zotero_write.py:364]: What do the line numbers refer too?

File: aiagents4pharma/talk2scholars/tests/test_zotero_write.py [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+    def test_zotero_read_string_match(
+        self,
+        mock_get_item_collections,
+        mock_zotero_class,
+        mock_hydra_compose,
+        mock_hydra_init,
+    ):
+        """
+        Test that if an entry in zotero_read is a string that matches the normalized path,
+        it is used as the collection key (lines 119-123).
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tests/test_zotero_write.py:420]: Also these line numbers?

File: aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py [hunk 4]
```diff
@@ -84,48 +84,50 @@ def get_single_paper_recommendations(
     if year:
         params["year"] = year
 
-    response = requests.get(endpoint, params=params, timeout=cfg.request_timeout)
-    data = response.json()
-    response = requests.get(endpoint, params=params, timeout=10)
-    # print(f"API Response Status: {response.status_code}")
-    logging.info(
+    # Wrap API call in try/except to catch connectivity issues and check response format
+    try:
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py:88]: Good idea with the exception handling to cover API errors, and integrating the logging to log the error.

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+
+class ZoteroSaveInput(BaseModel):
+    """Input schema for the Zotero save tool."""
+
+    tool_call_id: Annotated[str, InjectedToolCallId]
+    collection_path: str = Field(
+        default=None,
+        description=(
+            "The path where the paper should be saved in the Zotero library. "
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:35]:
```suggestion
            "The path where the paper should be saved in the Zotero library."
```

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            "Example: '/machine/cern/mobile'."
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:36]:
```suggestion
            "Example: '/machine/cern/mobile'"
```

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+            for path in paths:
+                if path.lower() == normalized_path:
+                    matched_collection_key = key
+                    logger.info(
+                        "Found direct match in zotero_read: %s -> %s", path, key
+                    )
+                    break
+        elif isinstance(paths, str) and paths.lower() == normalized_path:
+            matched_collection_key = key
+            logger.info("Found direct match in zotero_read: %s -> %s", paths, key)
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:133]: is a `break` missing here?

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    logger.info("Found %d collections", len(collections))
+
+    # Normalize the requested collection path (remove trailing slash, lowercase for comparison)
+    normalized_path = collection_path.rstrip("/").lower()
+
+    # Find matching collection
+    matched_collection_key = None
+
+    # First, try to directly find the collection key in zotero_read data
+    for key, paths in zotero_read_data.items():
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:122]: Searching through python dictionaries is not always the fastest operations. I am a bit concerned that as the collection size grows, the matching logic here will become a bottleneck. Would it be possible to transform the dictionary representation to a Pandas DataFrame and use the built in filter functions? If time is a concern (I would like to begin testing and developing the use cases ASAP), we can also discuss addressing this in a subsequent issue. I should note that in general, the more we can rely on Pandas DataFrame objects for any large (or potentially large) data structures that can be mapped to a tabular format the better. We can utilize various CPU and GPU acceleration libraries with little additional code overhead to reduce the user response time.

File: app/frontend/streamlit_app_talk2scholars.py [hunk 9]
```diff
@@ -236,134 +301,157 @@ def _submit_feedback(user_response):
...
-                            st.session_state.messages.append(
-                                {
-                                    "type": "dataframe",
-                                    "content": df_papers,
-                                    "key": "dataframe_" + uniq_msg_id,
-                                    "tool_name": msg.name,
-                                }
-                            )
-                            st.empty()
+                    # # Add assistant response to chat history
```
Comment (dmccloskey) [inline app/frontend/streamlit_app_talk2scholars.py:356]: I am assuming these comments are intentional?

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The PR looks good 🤩. In particular, the thoroughness of the tests is much much appreciated 👍. 

I had a few comments on 1. consolidating the testing utility classes/functions and 2. migrating to Pandas DataFrame objects as much as possible for any filtering, and searching operations.

File: aiagents4pharma/talk2scholars/tests/test_zotero_write.py (hunk 2 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_zotero_write.py:364]: Sorry, this was for my understanding while I was writing the tests, I just removed it.

File: aiagents4pharma/talk2scholars/tests/test_zotero_write.py (hunk 3 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_zotero_write.py:420]: same here

File: aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py (hunk 4 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py:88]: Thank you☺️

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py (hunk 7 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:133]: Thank you for catching that! You’re absolutely right—without a break in the outer loop, the function would continue iterating over zotero_read_data even after finding a match.
I’ve added a break after the inner loop to ensure that once a match is found, the loop exits immediately

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py (hunk 8 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:122]: I completely agree with your point, and I appreciate your suggestion regarding Pandas for optimized searching. I'll test this approach locally, but I’d like to complete this improvement in a subsequent PR since my end-semester exams are coming up next week.  

I’ve also explored GPU-accelerated options like [NVIDIA’s cuDF](https://github.com/rapidsai/cudf) and [Polars](https://github.com/pola-rs/polars) (written in Rust), which could significantly speed up large-scale operations. In fact, @gurdeep330 and I had discussed this concern in the past, and your feedback reinforces that direction.

File: aiagents4pharma/talk2scholars/tests/test_main_agent.py
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_main_agent.py:1]: Thank you for your kind words! I really appreciate the feedback, especially on the thoroughness of the tests.
I completely agree with your suggestions on consolidating the testing utility functions and migrating to Pandas for filtering and searching. I’d like to cover these improvements in subsequent PRs, as I have my end-semester exams coming up next week and an exam tomorrow. I plan to revisit this after the 14th.
Thank you for your understanding! Looking forward to implementing these enhancements soon.

File: app/frontend/streamlit_app_talk2scholars.py (hunk 9 above)
Comment (ansh-info) [reply app/frontend/streamlit_app_talk2scholars.py:387]: Yes, this would be cover during the integration phase

File: aiagents4pharma/talk2scholars/tests/test_main_agent.py
[the remark this answers is not in the document]
Comment (dmccloskey) [reply aiagents4pharma/talk2scholars/tests/test_main_agent.py:1]: I just saw the issue for the migration to Pandas. Please add the issue for the consolidation of the testing classes/functions if you have not done so already and I just missed it.

File: aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py (hunk 8 above)
Comment (dmccloskey) [reply aiagents4pharma/talk2scholars/tools/zotero/zotero_write.py:122]: Great.

I would recommend cuDF 😉. We plan to add in the option to use cuDF (and CuGraph) in other parts of the repo.

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The changes look good 👍, and I think we are ready to merge. However, before merging I wanted to double check that the deletion of the notebook was not an accident as it seems unrelated to the review?

File: aiagents4pharma/talk2scholars/tests/test_main_agent.py
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tests/test_main_agent.py:1]: I have created an issue for it - issue: #136

File: aiagents4pharma/talk2scholars/tutorial.ipynb
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tutorial.ipynb:1]: I have just added issue #136 
Regarding the notebook. The notebook was uploaded by mistake unintentionally, but I have removed it since it was empty.

**Review by dmccloskey — APPROVED** [review]
